zcash_client_sqlite/src/wallet/init/migrations/ensure_orchard_ua_receiver.rs

```diff
@@ -0,0 +1,200 @@
+//! This migration ensures that an Orchard receiver exists in the wallet's default Unified address.
+use std::collections::HashSet;
+
+use rusqlite::named_params;
+use schemer_rusqlite::RusqliteMigration;
+use uuid::Uuid;
+
+use zcash_client_backend::keys::{
+    UnifiedAddressRequest, UnifiedFullViewingKey, UnifiedIncomingViewingKey,
+};
+use zcash_primitives::consensus;
+
+use super::orchard_received_notes;
+use crate::{wallet::init::WalletMigrationError, UA_ORCHARD, UA_TRANSPARENT};
+
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0x604349c7_5ce5_4768_bea6_12d106ccda93);
+
+pub(super) struct Migration<P> {
+    pub(super) params: P,
+}
+
+impl<P> schemer::Migration for Migration<P> {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        [orchard_received_notes::MIGRATION_ID].into_iter().collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that the wallet's default address contains an Orchard receiver."
+    }
+}
+
+impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction<'_>) -> Result<(), Self::Error> {
+        let mut get_accounts = transaction.prepare(
+            r#"
+            SELECT id, ufvk, uivk
+            FROM accounts
+            "#,
+        )?;
+
+        let mut update_address = transaction.prepare(
+            r#"UPDATE "addresses"
+               SET address = :address
+               WHERE account_id = :account_id
+               AND diversifier_index_be = :j
+            "#,
+        )?;
+
+        let mut accounts = get_accounts.query([])?;
+        while let Some(row) = accounts.next()? {
+            let account_id = row.get::<_, u32>("id")?;
+            let ufvk_str: Option<String> = row.get("ufvk")?;
+            let uivk = if let Some(ufvk_str) = ufvk_str {
+                UnifiedFullViewingKey::decode(&self.params, &ufvk_str[..])
+                    .map_err(|_| {
+                        WalletMigrationError::CorruptedData("Unable to decode UFVK".to_string())
+                    })?
+                    .to_unified_incoming_viewing_key()
+            } else {
+                let uivk_str: String = row.get("uivk")?;
+                UnifiedIncomingViewingKey::decode(&self.params, &uivk_str[..]).map_err(|_| {
+                    WalletMigrationError::CorruptedData("Unable to decode UIVK".to_string())
+                })?
+            };
+
+            let (default_addr, diversifier_index) = uivk.default_address(
+                UnifiedAddressRequest::unsafe_new(UA_ORCHARD, true, UA_TRANSPARENT),
+            )?;
```

Comment:
What is the rationale for only adding Orchard to the default address?

- `WalletWrite::get_next_available_address` was introduced in #632, at which point the behaviour was "generate a UA with all of the wallet's receiver types".
  - Released in 0.6.0 on 2022-11-12.
- #829 added receiver type selection to this API. After this PR, the caller would have needed to explicitly specify which receiver types they wanted.
  - Released in 0.11.0 on 2024-03-01, less than one month ago.

Given the above, I think the vast majority of addresses (if not _all_ of them) were generated assuming implicit addressing (i.e. the default address is not special in that regard), and we should add Orchard receivers to all of them.

We might also want to distinguish addresses generated before 0.11.0 from addresses generated after (i.e. distinguish them _here_), so we can keep track of which ones should continue to be bumped. Or maybe we just distinguish which addresses were generated with a _subset_ of the wallet's supported receivers, so that callers who just want "all of them" get them bumped as well in future, and only UAs created with a specific subset are left alone by future migrations.

Reply:
Hm, I weakly disagree; my expectation is that if people are generating multiple addresses, that they're also not reusing addresses and so we shouldn't retroactively change things; the new addresses being generated with Orchard receivers going forward would fix the problem.

Reply:
Hmmm, okay I'll buy this.

Reply:
If this ends up removing a Sapling receiver from the UA it's migrating from, isn't that bad?